### PR TITLE
[data] fix reading multiple parquet files with ragged ndarrays

### LINKED ISSUE #47960: [data] read_parquet doesn't work for multiple parquet files with ragged ndarrays
When reading from multiple parquet files, if each file contains a different-shaped ndarray for the same column. The following error will be raised: `pyarrow.lib.ArrowTypeError: Casting from 'extension<ray.data.arrow_variable_shaped_tensor<ArrowVariableShapedTensorType>>' to different extension type 'extension<ray.data.arrow_tensor<ArrowTensorType>>' not permitted. One can first cast to the storage type, then to the extension type.`.

This is because schema is [only inferred based on the first parquet file](https://github.com/ray-project/ray/blob/46a97b80575824f501649cb3d2b887e6cbd0d91f/python/ray/data/_internal/datasource/parquet_datasource.py#L233-L236). This makes all read tasks output arrow tables with the same schema, although the actual data shape is different. Then when concatenating these tables, Ray Data [doesn't know](https://github.com/ray-project/ray/blob/46a97b80575824f501649cb3d2b887e6cbd0d91f/python/ray/data/_internal/arrow_ops/transform_pyarrow.py#L127) it needs to convert the tables to `ArrowVariableShapedTensorType`.

Repro script:
```python
import numpy as np

import ray

N = 3
output_path = "/tmp/output"

ds = ray.data.range(N)


def map(row):
    id = row["id"] + 1
    row["data"] = np.zeros((id * 100, id * 100), dtype=np.int8)
    return row


ds.map(map).repartition(N).write_parquet(output_path)

ds2 = ray.data.read_parquet(output_path, override_num_blocks=1)
ds2.take_all()
```

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

PyArrow infers parquet schema only based on the first file. This will cause errors when reading multiple files with ragged ndarrays. 

This PR fixes this issue by not using the inferred schema for reading. 

## Related issue number
Fixes https://github.com/ray-project/ray/issues/47960

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 1]
```diff
@@ -230,6 +230,14 @@ def __init__(
         # duplicating the partition data, we disable PyArrow's partitioning.
         dataset_kwargs["partitioning"] = None
 
+        # `read_schema` is the schema object that will be used to perform
+        # read operations.
+        # It should be None, unless user has specified the schema or columns.
+        # We don't use the inferred schema for read, because the pyarrow only infers
+        # schema based on the first file. Thus, files with different schemas will end
```
Comment (scottjlee) [inline python/ray/data/_internal/datasource/parquet_datasource.py:237]: does that mean this is a pyarrow bug?

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 2]
```diff
@@ -240,24 +248,22 @@ def __init__(
...
 
         check_for_legacy_tensor_type(schema)
 
         if _block_udf is not None:
             # Try to infer dataset schema by passing dummy table through UDF.
             dummy_table = schema.empty_table()
             try:
-                inferred_schema = _block_udf(dummy_table).schema
-                inferred_schema = inferred_schema.with_metadata(schema.metadata)
+                schema = _block_udf(dummy_table).schema
```
Comment (scottjlee) [inline python/ray/data/_internal/datasource/parquet_datasource.py:259]: by overwriting the existing value for `schema` here, would we be losing `schema.metadata` which is used in the next line?

Comment (raulchen) [reply python/ray/data/_internal/datasource/parquet_datasource.py:259]: oh, nice catch

File: python/ray/data/_internal/datasource/parquet_datasource.py (hunk 1 above)
Comment (raulchen) [reply python/ray/data/_internal/datasource/parquet_datasource.py:237]: not sure it's a bug, oh they intended to do so to avoid high overheads

**Review by scottjlee — APPROVED** [review]

**Pull request merged**